Pull Request - Optimized Project

bibschan opened the pull request:
Enhancements:
- organized the folder structure (added src, dist, util folders)
- divided the app.ts file into separate files for each functionality (each class has its own file)
- added webpack to merge all the .js files and minify the code and outputs a bundle.js file. (avoid several HTTP requests for the multiple scripts created)

caroldarski commented on src/ValidateFunctions.ts:
```diff
@@ -0,0 +1,32 @@
+
+import { Valitadable } from "Validatable";
+
+export module ValidateFunctions {
```
I think we could remove those 2 methods from here and create the methods for checking other logic. I.e.:
```
export function isValueLengthValid(input: Valitadable.Validatable): boolean {
  const hasValidMaxLength = input.maxLength ? input.value.length < input.maxLength : true;
  const hasValidMinLength = input.minLength ? input.value.length > input.minLength : true;
  return hasValidMaxLength && hasValidMinLength;
}
```

Then you could get simplify the checking on the `Validatable.validate`

caroldarski commented on src/Validatable.ts:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  //validation function
+  export function Validate(validateInput: Validatable) {
```
what do you think about changing it to camelCase? 😄 (not a big deal 😛 )

bibschan replied:
good catch! changing it now 😄

bibschan commented on index.html:
```diff
@@ -4,9 +4,13 @@
+    <title>Project Manager</title>
     <link rel="stylesheet" href="original.css" />
-    <script src="dist/app.js" defer></script>
+    <!-- <script>
```
good one! I left a few unimportant comments behind, I will clean up the code a bit before committing again 😃

bibschan commented on src/Validatable.ts:
```diff
@@ -0,0 +1,64 @@
+import { ValidateFunctions } from "./ValidateFunctions";
+
+export module Valitadable {
+  export interface Validatable {
+    value: string | number;
+    required?: boolean;
+    minLength?: number;
+    maxLength?: number;
+    min?: number;
+    max?: number;
+  }
+
+  //validation function
+  export function Validate(validateInput: Validatable) {
+    //we set this to false as soon as any var fails the test
+    let isValid = true;
+
+    if (validateInput.required)
+      //if the required var is set and true (not undefined) then value must not be empty
+      //checks if value is empty after trimming the string
+      isValid = isValid && validateInput.value.toString().trim().length !== 0;
+    //if minLength is set then it should have a min length. only checks strings, doesnt make sense
+    //to check length if a number
+    //
+    //
+    //
+    //
+    //Carol, aqui está o problema, se eu uso as funções do ValidateFunctions, ele não
+    //consegue entender que .value.length é uma string, apesar de ter o check === "string"
+    //no validateFunctions.
+    //validação inicial
+    if (
+      validateInput.minLength != null &&
+      typeof validateInput.value === "string"
+    )
```
Working on it! Thanks for the feedback and I agree that Validatable can look cleaner, I'm thinking about keeping the Validatable interface as is and just performing the checks in separate files so it doesn't look too cluttered.

bibschan commented:
Hi @caroldarski! I worked on your pointers this morning and split the validation into two files, ValidateNumber and ValidateString, but kept the Validatable interface as is. The checks are now performed in those new files and Validatable looks a lot less cluttered. I also cleaned up some old comments that were left behind in the process 😄 Let me know what you think!

caroldarski commented on src/ProjectInput.ts:
```diff
@@ -0,0 +1,96 @@
+import { Component } from "Component";
+import { autobind } from "autobind";
+import { ProjectState } from "ProjectState";
+import { Valitadable } from "Validatable";
```
@bibschan some weird thing happened. it seems that it didn't get your last commit changes here 👀 I downloaded the code as well and I am still seeing the reference to the old Validatable. Could you please check?

caroldarski replied:
nvm! I think github was having a hard time, now seems to be everything updates 🤷‍♀️

caroldarski replied:
I think we can delete this file as it isn't being used anywhere, right? 😄

bibschan replied:
Hi @caroldarski - ProjectInput is where the app grabs the HTML elements and renders the input data, those import statements are also needed! Were you referring to another file perhaps?

caroldarski replied:
@bibschan I think the one being used by `app.ts` is the `"./components/ProjectInput"` this one is inside of `/src`
Is there anywhere else where this would be used? 😄 Otherwise, I believe you can delete it

bibschan replied:
OH YES! I finally found it - there's an inconsistency between my local repo and the GitHub one, my computer's Terminal link to Git is broken so I've been uploading manually. It's fixed now! 😄

caroldarski commented on src/util/Validatable.ts:
```diff
@@ -0,0 +1,28 @@
+  //validation function
+  export function validate(validateInput: Validatable) {
+    let isValid = true; //we set this to false as soon as any variable fails the test
+    isValid = ValidateString.requiredCheck(isValid, validateInput);
+    isValid = ValidateString.maxLengthCheck(isValid, validateInput);
+    isValid = ValidateString.minLengthCheck(isValid, validateInput);
+    isValid = ValidateNumber.minValueCheck(isValid, validateInput);
+    isValid = ValidateNumber.maxValueCheck(isValid, validateInput);
```
I think you need to append the isValid here, otherwise if the `requiredCheck` returns `false` and the `maxLengthCheck` returns true with all the others checks, it will give a false positive.

caroldarski replied:
Also, I would check the type here, something like this:
```
if (typeof validateInput.value === "string") {
  isValid = isValid && ValidateString.requiredCheck(isValid, validateInput);
  isValid = isValid && ValidateString.maxLengthCheck(isValid, validateInput);
  isValid = ValidateString.minLengthCheck(isValid, validateInput);
} else if (typeof validateInput.value === "number") {
...
}
```

bibschan replied:
WOW thanks for catching that! Now looking at it makes total sense to append the isValid variable 😸

caroldarski replied:
@bibschan I think this file still not updated with the logic, could you please check?

caroldarski replied:
it seems you have duplicated files here as well: this one is inside of `src/util/Validatable.ts` and the most recent one is under `util/Validatable.ts`

bibschan replied:
I was finally able to fix those inconsistencies! thanks for the feedback <3 let me know if you find anything else

caroldarski commented on src/util/ValidateString.ts:
```diff
@@ -0,0 +1,40 @@
+    isValid: boolean,
+    validateInput: Validatable.Validatable
+  ) {
+    if (
+      validateInput.minLength != null &&
+      typeof validateInput.value === "string"
+    )
```
> Also, I would check the type here, something like this:
> 
> ```
> if (typeof validateInput.value === "string") {
>   isValid = isValid && ValidateString.requiredCheck(isValid, validateInput);
>   isValid = isValid && ValidateString.maxLengthCheck(isValid, validateInput);
>   isValid = isValid && ValidateString.minLengthCheck(isValid, validateInput);
> } else if (typeof validateInput.value === "number") {
> ...
> }
> ```

if you're going to check the String outside, here you don't need to check again, leaving the code cleaner ✨  (same for all the functions here and on the ValidateNumber)

bibschan replied:
Gotcha! I'll stick to the ValidateString and ValidateNumber checks then 😃

bibschan commented:
Fixed the logic issue in Validatable (appended isValid variable to all checks).
